core/drivers/atmel_shdwc_a32.S

```diff
@@ -0,0 +1,59 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2015 Atmel Corporation,
+ *                    Nicolas Ferre <nicolas.ferre@atmel.com>
+ * Copyright (c) 2021, Microchip
+ */
+
+
+#include <arm.h>
+#include <arm32_macros.S>
+#include <asm.S>
+#include <at91_pmc.h>
+#include <drivers/atmel_shdwc.h>
+#include <drivers/sam/at91_ddr.h>
+
+/**
+ * Shutdown the CPU
+ *
+ * This function is in assembly to be sure the code fits in a single cache line.
+ * We are going to power down the SDRAM and thus we can't fetch code from it
```

Comment:
So you're depending on the cache line staying in the icache? Any guarantees for that?

Reply:
This was the intended purpose of this block of code taken from Linux sama5d2 shutdown controller. But after looking at it carefully, it probably does not work as-is.
The cache line on Cortex A5 is 8 words long (32 bytes) The actual code is 40 bytes long. So there is no guarantee that the code will fit in a single cache line. The code was originally working but ended up being modified non working IMHO.
I would say that if the code is aligned on 32 bytes and is smaller than 8 bytes, there is no reason for it to be evicted (unless there are some exceptions happening which as I say will be disabled) . I could also add a compile time check to be sure the code is indeed less than 32 bytes.

In order to make this code working, I will modified it to align on a cache line just before powering the SDRAM (`str	r3, [r0, #AT91_DDRSDRC_LPR]`), this result in the end of the code being 8 bytes long (minus the infinite loop). But either the SDRAM is not powered down and the infinite loop will be taken, or the SDRAM is powered down and this line will probably never be loaded from SDRAM and the processor will stall.

Reply:
What if there's an icache invalidate from another CPU as part of some non-secure attack? Can that cause the cache line to be removed?
Perhaps that's not a security issue, but only a DOS.

Reply:
OK, thanks.